ipa-core/src/bin/report_collector.rs

```diff
@@ -286,6 +310,93 @@ async fn main() -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
+fn inputs_from_url_file(
+    url_file_path: &Path,
+    query_id: QueryId,
+    shard_count: usize,
+) -> Result<Vec<[QueryInput; 3]>, Box<dyn Error>> {
+    let mut file = BufReader::new(File::open(url_file_path)?);
+    let mut buf = String::new();
+    let mut inputs = [Vec::new(), Vec::new(), Vec::new()];
+    for (helper_id, helper_input) in inputs.iter_mut().enumerate() {
+        for _ in 0..shard_count {
+            buf.clear();
+            if file.read_line(&mut buf)? == 0 {
+                break;
+            }
+            helper_input
+                .push(Uri::try_from(buf.trim()).map_err(|e| format!("Invalid URL {buf:?}: {e}"))?);
+        }
+        if helper_input.len() != shard_count {
+            return Err(format!(
+                "Helper {helper_id} does not have enough input. Expected {shard_count}, got {}",
```

Comment:
This is a nit given the nature of this exercise, but: maybe just read all the lines immediately and complain if the file doesn't have the expected number of lines (3 * N)?

Attributing the incorrect length to a particular helper may be confusing, and this will accept a file that is too long, including e.g. if the length is 3 * M for an incorrect number of shards M > N.